### PR TITLE
qemu.test: Update case "readonly_disk"

### PR DESCRIPTION
**Pull request by XueqiangWei**
Signed-off-by: Xueqiang Wei <xuwei@redhat.com>
ID: 1533394 1483858

### CODE REVIEW

**Comment by XueqiangWei** [comment]
Test results (q35 + pc): 
(1/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.q35.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.readonly_disk: PASS (827.04 s)
 (2/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.q35.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror: PASS (206.44 s)

(1/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.readonly_disk: PASS (787.71 s)
 (2/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror: PASS (224.01 s)

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by CongLi** [comment]
@luckyh 

Could you please help review this PR?

Thanks.

**Comment by yanan-fu** [comment]
NACK.
The reason is same with https://github.com/autotest/tp-qemu/pull/928.

**Comment by CongLi** [comment]
Thank @yanan-fu .

@luckyh 
Could you please provide more details of this issue?

**Comment by XueqiangWei** [comment]
depends on avocado-framework/avocado-vt#1457

**Comment by XueqiangWei** [comment]
(1/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.q35: PASS (223.48 s)
 (2/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror.q35: PASS (219.78 s)
 (1/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk: PASS (232.25 s)
 (2/2) smp_4.2048m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror: PASS (215.31 s)

**Comment by XueqiangWei** [comment]
depends on avocado-framework/avocado-vt#1556

**Comment by XueqiangWei** [comment]
(1/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk: PASS (282.15 s)
 (2/2) smp_8.8192m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror: PASS (282.85 s)

**Comment by XueqiangWei** [comment]
depends on avocado-framework/avocado-vt#1640

**Comment by XueqiangWei** [comment]
(1/2) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk: PASS (250.10 s)
(2/2) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror: PASS (209.92 s)

**Comment by XueqiangWei** [comment]
@hereischen @YongxueHong @aliang123 @lolyu  please help review it. Many thanks.

**Comment by XueqiangWei** [comment]
Tested with python3 and windows2019, the results are passed.

File: qemu/tests/readonly_disk.py [hunk 1]
```diff
@@ -32,16 +31,33 @@ def run(test, params, env):
         "TEST STEPS 2: Format the disk and copy file to it", logging.info)
     os_type = params["os_type"]
     copy_cmd = params.get("copy_cmd", "copy %s %s")
-    disk_idx = params.get("disk_index", 1)
-    fs_type = params.get("fstype", "ntfs")
-    drive_letter = params.get("drive_letter", "I:")
-    disk_size = params.get("partition_size_data", "200M")
+    fstype = params.get("fstype", "ntfs")
+    data_image_size = params.get("image_size_data", "1G")
+    data_image_num = int(params.get("data_image_num",
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: seems no need default value here if only test with 1 data disk.

File: qemu/tests/readonly_disk.py [hunk 2]
```diff
@@ -32,16 +31,33 @@ def run(test, params, env):
...
+                                    len(params.objects("images")) - 1))
+    err_msg = "Set disks num: %d" % data_image_num
+    err_msg += ", Get disks num in guest: %d"
+    error_context.context("Get windows disk index that to "
+                          "be formatted", logging.info)
+    disks = utils_disk.get_windows_disks_index(session, data_image_size)
+    if len(disks) < data_image_num:
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: for this case ,the disks list should only has 1 cell, why not condition if len(disks) == 0? can reduce param data_image_num

File: qemu/tests/readonly_disk.py [hunk 3]
```diff
@@ -32,16 +31,33 @@ def run(test, params, env):
...
+        test.fail("Fail to list all the volumes"
+                  ", %s" % err_msg % len(disks))
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: suggest change to follows and move line 38, 39
test.fail("Fail to list all the volumes,"
             "Set disks num: %d ," 
             "Get disks num in guest: %d" % ( data_image_num, len(disks)))

File: qemu/tests/readonly_disk.py [hunk 4]
```diff
@@ -32,16 +31,33 @@ def run(test, params, env):
...
         session, params["src_file"], label="WIN_UTILS")
-    utils_misc.format_guest_disk(session, disk_idx, drive_letter,
-                                 disk_size, fs_type, os_type)
-    dst_file = params["dst_file"]
-    session.cmd(copy_cmd % (src_file, dst_file))
+    error_context.context("Clear readonly for all disks and online "
+                          "them in guest.", logging.info)
+    if not utils_disk.update_windows_disk_attributes(session, disks):
+        test.fail("Failed to update windows disk attributes.")
+    for disk in disks:
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: Confused why not use disk_index directly from cfg? it's more Simplify the code, will no need use a loop here~

File: qemu/tests/readonly_disk.py (hunk 1 above)
Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: It is ok, can be used universally.

File: qemu/tests/readonly_disk.py (hunk 2 above)
Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: this condition used universally

File: qemu/tests/readonly_disk.py (hunk 3 above)
Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: I think they are all ok. It is too long that all in a line.

File: qemu/tests/readonly_disk.py (hunk 4 above)
Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: We need  to think about the code used universally.

Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: disks value is a list.  use \'for\" loop to navigate it.

File: qemu/tests/readonly_disk.py [hunk 5]
```diff
@@ -32,16 +31,33 @@ def run(test, params, env):
...
+        error_context.context("Format disk in guest: '%s'" % disk,
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: it's better change to error_context.context("Format disk %s in guest" % disk, logging.info) .
currently code will print info as " Format disk in guest: '1' ", user will confuse the meaning for end number '1'.

Comment (XueqiangWei) [reply qemu/tests/readonly_disk.py]: disk index mentioned in line 40, error_context.context(\"Get windows disk index that to be formatted\", logging.info)

Comment (peixiu) [reply qemu/tests/readonly_disk.py]: Althrough disk index mentioned in line 40, but like this comment "01:43:41 INFO | Context: Format disk in guest: '1'",  I thinks it's so confusing user if not check the code source~

**Comment by XueqiangWei** [comment]
(1/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.q35: PASS (314.95 s)
 (2/2) Host_RHEL.m7.u6.qcow2.virtio_scsi.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.readonly_disk.with_wrerror.q35: PASS (383.52 s)

Tested with python3, all are passed.

**Comment by hereischen** [comment]
Acked-by: Haotong Chen hachen@redhat.com

File: qemu/tests/readonly_disk.py [hunk 6]
```diff
@@ -60,16 +76,15 @@ def run(test, params, env):
     error_context.context(
         "TEST STEPS 4: Write to the readonly disk expect:"
         "The media is write protected", logging.info)
-    dst_file_readonly = params["dst_file_readonly"]
+    dst_file_readonly = params["dst_file_readonly"] % drive_letter[0]
```
Comment (peixiu) [inline qemu/tests/readonly_disk.py]: According to 52 line,  if disks have 2 or more values, here  drive_letter[0] only get the value for last time assignment, it should be assigned value for each disk~

**Comment by peixiu** [comment]
Acked-by: Peixiu Hou phou@redhat.com

**Comment by YongxueHong** [comment]
Thanks all.

**Pull request merged**